File: src/morphic/frontend/routes_dupfinder.py
"""
Dupfinder-tab API routes — scan, status, results, delete.
"""

from __future__ import annotations

import os
import time

from flask import Blueprint, jsonify, request

from morphic.shared.constants import (
    DEFAULT_IMAGE_THRESHOLD,
    DEFAULT_VIDEO_THRESHOLD,
)
from morphic.shared.utils import format_file_size
from morphic.dupfinder.scanner import get_job, start_job

bp = Blueprint("dupfinder", __name__)


# ── Scan ───────────────────────────────────────────────────────────────────


@bp.route("/scan", methods=["POST"])
def api_start_scan():
    """Start a new duplicate scan."""
    data = request.get_json()
    if not data:
        return jsonify({"error": "No JSON body"}), 400

    folder = data.get("folder", "").strip()
    scan_type = data.get("type", "both")
    image_threshold = float(
        data.get("image_threshold", DEFAULT_IMAGE_THRESHOLD),
    )
    video_threshold = float(
        data.get("video_threshold", DEFAULT_VIDEO_THRESHOLD),
    )

    if not folder or not os.path.isdir(folder):
        return jsonify({"error": f"Invalid folder: {folder}"}), 400
    if scan_type not in ("images", "videos", "both"):
        return jsonify(
            {
                "error": "type must be images, videos, or both",
            }
        ), 400

    job_id = start_job(
        folder=folder,
        scan_type=scan_type,
        image_threshold=image_threshold,
        video_threshold=video_threshold,
    )
    return jsonify({"job_id": job_id}), 202


# ── Status & results ──────────────────────────────────────────────────────


@bp.route("/scan/<job_id>/status")
def api_scan_status(job_id: str):
    """Get status of a scan job."""
    job = get_job(job_id)
    if not job:
        return jsonify({"error": "Job not found"}), 404

    elapsed = 0.0
    if job.started_at:
        end = job.finished_at if job.finished_at else time.time()
        elapsed = end - job.started_at

    return jsonify(
        {
            "id": job.id,
            "status": job.status,
            "progress": job.progress,
            "message": job.message,
            "error": job.error,
            "total_files_found": job.total_files_found,
            "total_files_processed": job.total_files_processed,
            "elapsed_seconds": round(elapsed, 1),
        }
    )


@bp.route("/scan/<job_id>/results")
def api_scan_results(job_id: str):
    """Get results of a completed scan."""
    job = get_job(job_id)
    if not job:
        return jsonify({"error": "Job not found"}), 404

    if job.status not in ("done", "error"):
        return jsonify({"error": "Scan not finished yet"}), 409

    return jsonify(
        {
            "image_groups": job.image_groups,
            "video_groups": job.video_groups,
            "space_savings": job.space_savings,
            "space_savings_formatted": format_file_size(job.space_savings),
        }
    )


# ── Delete ─────────────────────────────────────────────────────────────────


@bp.route("/delete", methods=["POST"])
def api_delete_files():
    """Delete selected duplicate files."""
    data = request.get_json()
    if not data:
        return jsonify({"error": "No JSON body"}), 400

    files = data.get("files", [])
    if not files:
        return jsonify({"error": "No files specified"}), 400

    results = []
    total_freed = 0

    for file_path in files:
        try:
            if not os.path.isfile(file_path):
                results.append({"path": file_path, "status": "not_found"})
                continue
            file_size = os.path.getsize(file_path)
            os.remove(file_path)
            total_freed += file_size
            results.append(
                {
                    "path": file_path,
                    "status": "deleted",
                    "size_freed": file_size,
                }
            )
        except PermissionError:
            results.append(
                {
                    "path": file_path,
                    "status": "permission_denied",
                }
            )
        except Exception as e:
            results.append(
                {
                    "path": file_path,
                    "status": "error",
                    "error": str(e),
                }
            )

    return jsonify(
        {
            "results": results,
            "total_freed": total_freed,
            "total_freed_formatted": format_file_size(total_freed),
        }
    )

File: web/routes_dupfinder.go
package web

import (
	"net/http"
	"os"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/exterex/morphic/internal/dupfinder"
	"github.com/exterex/morphic/internal/shared"
)

func registerDupfinderRoutes(r *gin.Engine) {
	g := r.Group("/api/dupfinder")
	{
		g.POST("/scan", handleDupfinderScan)
		g.GET("/scan/:id/status", handleDupfinderStatus)
		g.GET("/scan/:id/results", handleDupfinderResults)
		g.POST("/delete", handleDupfinderDelete)
	}
}

func handleDupfinderScan(c *gin.Context) {
	var req struct {
		Folder         string  `json:"folder"`
		Type           string  `json:"type"`
		ImageThreshold float64 `json:"image_threshold"`
		VideoThreshold float64 `json:"video_threshold"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if req.Folder == "" || !isDir(req.Folder) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid folder: " + req.Folder})
		return
	}
	if req.Type == "" {
		req.Type = "both"
	}
	if req.Type != "images" && req.Type != "videos" && req.Type != "both" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "type must be images, videos, or both"})
		return
	}
	if req.ImageThreshold == 0 {
		req.ImageThreshold = shared.DefaultImageThreshold
	}
	if req.VideoThreshold == 0 {
		req.VideoThreshold = shared.DefaultVideoThreshold
	}

	jobID := dupfinder.StartJob(req.Folder, req.Type, req.ImageThreshold, req.VideoThreshold)
	c.JSON(http.StatusAccepted, gin.H{"job_id": jobID})
}

func handleDupfinderStatus(c *gin.Context) {
	id := c.Param("id")
	job, ok := dupfinder.GetJob(id)
	if !ok {
		c.JSON(http.StatusNotFound, gin.H{"error": "Job not found"})
		return
	}

	elapsed := 0.0
	if !job.StartedAt.IsZero() {
		end := job.DoneAt
		if end.IsZero() {
			end = time.Now()
		}
		elapsed = end.Sub(job.StartedAt).Seconds()
	}

	c.JSON(http.StatusOK, gin.H{
		"id":                   job.ID,
		"status":               job.Status,
		"progress":             job.Progress,
		"message":              job.Message,
		"error":                job.Error,
		"total_files_found":     job.TotalFound,
		"total_files_processed": job.TotalProcessed,
		"elapsed_seconds":      round1(elapsed),
	})
}

func handleDupfinderResults(c *gin.Context) {
	id := c.Param("id")
	job, ok := dupfinder.GetJob(id)
	if !ok {
		c.JSON(http.StatusNotFound, gin.H{"error": "Job not found"})
		return
	}

	if job.Status != "done" && job.Status != "failed" {
		c.JSON(http.StatusConflict, gin.H{"error": "Scan not finished yet"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"image_groups":          job.ImageGroups,
		"video_groups":          job.VideoGroups,
		"space_savings":         job.SpaceSavings,
		"space_savings_formatted": shared.FormatFileSize(job.SpaceSavings),
	})
}

func handleDupfinderDelete(c *gin.Context) {
	var req struct {
		Files []string `json:"files"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if len(req.Files) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "No files specified"})
		return
	}

	var results []map[string]interface{}
	totalFreed := int64(0)

	for _, fp := range req.Files {
		info, err := os.Stat(fp)
		if err != nil {
			results = append(results, map[string]interface{}{"path": fp, "status": "not_found"})
			continue
		}
		if info.IsDir() {
			results = append(results, map[string]interface{}{"path": fp, "status": "not_found"})
			continue
		}
		size := info.Size()
		if err := os.Remove(fp); err != nil {
			if os.IsPermission(err) {
				results = append(results, map[string]interface{}{"path": fp, "status": "permission_denied"})
			} else {
				results = append(results, map[string]interface{}{"path": fp, "status": "error", "error": err.Error()})
			}
		} else {
			totalFreed += size
			results = append(results, map[string]interface{}{"path": fp, "status": "deleted", "size_freed": size})
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"results":              results,
		"total_freed":          totalFreed,
		"total_freed_formatted": shared.FormatFileSize(totalFreed),
	})
}
